### ISSUE TITLE
OCI DB System patching precheck and apply operation not working

### ISSUE THREAD

**Issue by rishimahajan** [oracle/oci-ansible-modules#86]
## Issue Report

**Describe the issue**
Trying to run a PRECHECK and APPLY operation on a VM DB system.  Both operations are working from OCI console but i couldn't make it work via ansible modules.

**Expected behavior**
A precheck operation should run for 'PRECHECK' action and
an apply operation should run for 'APPLY' operation.

**Environment**
OCI

**OS version:**
Oracle Linux Server release 7.7

**Ansible version:**
Tried with 2 versions
ansible 2.7.7
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/opc/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Jun 20 2019, 13:30:36) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36.0.1)]

ansible 2.8.4
  config file = /home/opc/dbpatching/ansible.cfg
  configured module search path = [u'/home/opc/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Aug  7 2019, 08:19:52) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39.0.1)]

**OCI Python SDK version:**
python -c "import oci;print(oci.__version__)"
2.6.0

**OCI Ansible Modules version:**

python -c "from ansible.module_utils.oracle import oci_utils; print(oci_utils.__version__)"
1.12.0

**Ansible playbook to reproduce the issue**
```
- name: Patch a DB System
  hosts: localhost
  vars:
    db_ocid: ocid1.dbsystem.oc1.iad.xxxxxxx
  tasks:
    - name: PRECHECK a patch on the DB System
      oci_db_system:
        db_system_id: "ocid1.dbsystem.oc1.iad.xxxxx"
        version:
          patch_id: "ocid1.dbpatch.oc1.iad.xxxx"
          action: 'PRECHECK'
        state: 'present'
      register: precheck
...
[opc@tower dbpatching]$ OCI_ANSIBLE_AUTH_TYPE="instance_principal" ansible-playbook check.yml -vvv
TASK [PRECHECK a patch on the DB System] *********************************************************************************************************************
task path: /home/opc/dbpatching/check.yml:8
Wednesday 20 November 2019  05:07:51 +0000 (0:00:00.916)       0:00:01.070 **** 
Using module file /usr/lib/python2.7/site-packages/ansible/modules/cloud/oracle/oci_db_system.py
Pipelining is enabled.
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: opc
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python2 && sleep 0'
The full traceback is:
WARNING: The below traceback may *not* be related to the actual failure.
  File "/tmp/ansible_oci_db_system_payload_VU326H/ansible_oci_db_system_payload.zip/ansible/module_utils/oracle/oci_utils.py", line 738, in create_resource
    resource = to_dict(call_with_backoff(create_fn, **kwargs_create).data)
  File "/tmp/ansible_oci_db_system_payload_VU326H/ansible_oci_db_system_payload.zip/ansible/module_utils/oracle/oci_utils.py", line 1694, in call_with_backoff
    return fn(**kwargs)
  File "/usr/lib/python2.7/site-packages/oci/database/database_client.py", line 9921, in update_db_system
    response_type="DbSystem")
  File "/usr/lib/python2.7/site-packages/oci/retry/retry.py", line 271, in make_retrying_call
    return func_ref(*func_args, **func_kwargs)
  File "/usr/lib/python2.7/site-packages/oci/base_client.py", line 243, in call_api
    return self.request(request)
  File "/usr/lib/python2.7/site-packages/oci/base_client.py", line 358, in request
    self.raise_service_error(request, response)
  File "/usr/lib/python2.7/site-packages/oci/base_client.py", line 528, in raise_service_error
    original_request=request)

fatal: [localhost]: FAILED! => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "api_user": null, 
            "api_user_fingerprint": null, 
            "api_user_key_file": null, 
            "api_user_key_pass_phrase": null, 
            "auth_type": "api_key", 
            "availability_domain": null, 
            "backup_subnet_id": null, 
            "cluster_name": null, 
            "compartment_id": null, 
            "config_file_location": null, 
            "config_profile_name": null, 
            "cpu_core_count": null, 
            "data_storage_percentage": null, 
            "data_storage_size_in_gbs": null, 
            "database_edition": null, 
            "db_home": null, 
            "db_system_id": "ocid1.dbsystem.oc1.iad.abuwcljtbdjwmubo6v5r3qzni2shjyr2vgpl3vzllwkd3buj4gjfzhiuxbgq", 
            "defined_tags": null, 
            "delete_ssh_public_keys": false, 
            "disk_redundancy": null, 
            "display_name": null, 
            "domain": null, 
            "force_create": false, 
            "freeform_tags": null, 
            "hostname": null, 
            "initial_data_storage_size_in_gb": null, 
            "key_by": null, 
            "license_model": null, 
            "node_count": null, 
            "purge_ssh_public_keys": true, 
            "region": null, 
            "shape": null, 
            "source": "NONE", 
            "ssh_public_keys": null, 
            "state": "present", 
            "subnet_id": null, 
            "tenancy": null, 
            "version": {
                "action": "PRECHECK", 
                "patch_id": "ocid1.dbpatch.oc1.iad.abuwcljtix63d77k4omkwfnx6iqqf5xhkzjkzzeg5yfubsup2g42suyyy7zq"
            }, 
            "wait": true, 
            "wait_timeout": 1200, 
            "wait_until": null
        }
    }, 
    "msg": "{'status': 400, 'message': u'The minimum Data storage size (in GB) for upgrade is 2048.', 'code': u'InvalidParameter', 'opc-request-id': '308D927BD9174E8E8BEB722EA020D98A/41ECF4EC4A139843DD9BF5B5C7983258/954912480032193ADFC328222D1623B2'}"
}

PLAY RECAP ***************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   

Wednesday 20 November 2019  05:07:53 +0000 (0:00:01.887)       0:00:02.957 **** 
=============================================================================== 
PRECHECK a patch on the DB System --------------------------------------------------------------------------------------------------------------------- 1.89s
/home/opc/dbpatching/check.yml:8 ----------------------------------------------------------------------------------------------------------------------------
Gathering Facts --------------------------------------------------------------------------------------------------------------------------------------- 0.92s
/home/opc/dbpatching/check.yml:3 ----------------------------------------------------------------------------------------------------------------------------
```

**Comment by nalsaber** [comment]
This is a server error:
 "msg": "{'status': 400, 'message': u'The minimum Data storage size (in GB) for upgrade is 2048.', 'code': u'InvalidParameter', 'opc-request-id': '308D927BD9174E8E8BEB722EA020D98A/41ECF4EC4A139843DD9BF5B5C7983258/954912480032193ADFC328222D1623B2'}"

I would provide this parameter in my oci_db_system task:
    data_storage_size_in_gbs: description: Size, in GBs, to which the currently attached storage needs to be scaled up to for VM based DB system. This must be greater than current storage size. Note that the total storage size attached will be more than what is requested, to account for REDO/RECO space and software volume. This option required only for update operation.

**Comment by rishimahajan** [comment]
I believe the message is misleading. I am trying to run precheck for patching, not a storage  scale up. I tried the same operation for systems with 256, 512 and 2048 GB size and every-time i got similar message. This is my yml file.

- name: Patch a DB System
  hosts: localhost
  vars:
    db_ocid: ocid1.dbsystem.oc1.iad.xxxxxxx
  tasks:
    - name: PRECHECK a patch on the DB System
      oci_db_system:
        db_system_id: "ocid1.dbsystem.oc1.iad.xxxxx"
        version:
          patch_id: "ocid1.dbpatch.oc1.iad.xxxx"
          action: 'PRECHECK'
        state: 'present'
      register: precheck

**Comment by nalsaber** [comment]
Can you try it using CLI command? Here is an example:
https://docs.cloud.oracle.com/iaas/tools/oci-cli/latest/oci_cli_docs/cmdref/db/system/patch.html
oci db system patch  --database-id=$dbid --patch-id=$dbpatchid --patch-action=PRECHECK

**Comment by rishimahajan** [comment]
I have already tried and opened an issue for CLI also.
https://github.com/oracle/oci-cli/issues/223
With CLI, i am not getting any error message but precheck is not getting invoked on the database.

**Comment by nalsaber** [comment]
This seems a service issue. Let's verify this feature using CLI first.

**Comment by rishimahajan** [comment]
This doesn't look like a service issue. Tried with rest API multiple times and and it worked fine without any failures

[opc@ebsbasash11 ~]$ cat patch.sh 
. ./oci-curl
oci-curl database.us-ashburn-1.oraclecloud.com PUT ./patch.json "/20160918/dbSystems/ocid1.dbsystem.oc1.iad.abuwcljtci6jjvnfo3afhawletcbavbzbm4bmub575hq443vcvfoduzm2jva"
[opc@ebsbasash11 ~]$ ./patch.sh 
{
  "availabilityDomain" : "aBLb:US-ASHBURN-AD-1",
  "backupNetworkNsgIds" : null,
  "backupSubnetId" : null,
  "clusterName" : null,
  "compartmentId" : "ocid1.compartment.oc1..aaaaaaaaf6tcpns5nwjmdwl7chw4bgjwl7x7xskvipcchplw67fbiy3blazq",
  "cpuCoreCount" : 2,
  "dataStoragePercentage" : 80,
  "dataStorageSizeInGBs" : 256,
  "databaseEdition" : "ENTERPRISE_EDITION_HIGH_PERFORMANCE",
  "dbSystemOptions" : {
    "storageManagement" : "ASM"
  },
  "definedTags" : { },
  "diskRedundancy" : "HIGH",
  "displayName" : "pythdb",
  "domain" : "bassubad1.ebsvcn.oraclevcn.com",
  "faultDomains" : [ "FAULT-DOMAIN-1" ],
  "freeformTags" : { },
  "hostname" : "pythdb",
  "id" : "ocid1.dbsystem.oc1.iad.abuwcljtci6jjvnfo3afhawletcbavbzbm4bmub575hq443vcvfoduzm2jva",
  "iormConfigCache" : null,
  "kmsKeyId" : null,
  "lastMaintenanceRunId" : null,
  "lastPatchHistoryEntryId" : "ocid1.dbpatchhistory.oc1.iad.abuwcljtyyp5aboyn4wmon3o6lrux74quddvimnygl5onbit5mtaf3xexlma",
  "licenseModel" : "LICENSE_INCLUDED",
  "lifecycleDetails" : null,
  "lifecycleState" : "UPDATING",
  "listenerPort" : 1521,
  "maintenanceWindow" : null,
  "nextMaintenanceRunId" : null,
  "nodeCount" : 1,
  "nsgIds" : null,
  "recoStorageSizeInGB" : 256,
  "scanDnsRecordId" : "ocid1.vcndnsrecord.oc1.iad.abuwcljtjellhu2lz3wqclp5yrfhhsz2npbqbor7xfza7ih6zhsjjarfl6rq",
  "scanIpIds" : null,
  "shape" : "VM.Standard2.2",
  "sparseDiskgroup" : null,
  "sshPublicKeys" : [ "ssh-rsa xxxxxx" ],
  "subnetId" : "ocid1.subnet.oc1.iad.aaaaaaaacz6gt36l77zxt6gcleq5ojfax4t7sd6wi4i2ktvhz2wgt7xubbdq",
  "timeCreated" : "2019-07-25T02:53:14.314Z",
  "timeZone" : "UTC",
  "version" : "18.6.0.0.190416",
  "vipIds" : null
}[opc@ebsbasash11 ~]$ 
[opc@ebsbasash11 ~]$

**Comment by nalsaber** [comment]
I ran the DB system precheck and it is working without errors if you provide the size requested by the service error (2048). We need to check with the service team why are we getting the size error.

Here are the details:
- name: Update DB System Version PRECHECK
  oci_db_system:
    db_system_id: "{{ db_system_id }}"
    version: 
      patch_id: "{{ patch_id }}"
      action: 'PRECHECK'
    data_storage_size_in_gbs: 2048
    state: 'present'
    wait: False
  register: result


changed: [testhost] => {
    "changed": true,
    "db_system": {
        "availability_domain": "IwGV:US-ASHBURN-AD-2",
        "backup_network_nsg_ids": null,
        "backup_subnet_id": null,
        "cluster_name": "db-clus-793",
        "compartment_id": "ocid1.compartment.oc1....",
        "cpu_core_count": 1,
        "data_storage_percentage": 80,
        "data_storage_size_in_gbs": 1024,
        "database_edition": "ENTERPRISE_EDITION_EXTREME_PERFORMANCE",
        "defined_tags": {
            "ansible_tag_namespace_integration_test_1": {
                "ansible_tag_1": "small"
            }
        },
        "disk_redundancy": "NORMAL",
        "display_name": "ansible-db-system-793",
        "domain": "ansibledbvcn793.ansibledbvcn793.oraclevcn.com",
        "fault_domains": [
            "FAULT-DOMAIN-1"
        ],
        "freeform_tags": {
            "system_license": "trial"
        },
        "hostname": "db-system-793",
        "id": "ocid1.dbsystem.oc1.iad....a",
        "iorm_config_cache": null,
        "last_patch_history_entry_id": "ocid1.dbpatchhistory.oc1.iad....a",
        "license_model": "LICENSE_INCLUDED",
        "lifecycle_details": null,
        "lifecycle_state": "UPDATING",
        "listener_port": 1521,
        "node_count": 1,
        "nsg_ids": null,
        "reco_storage_size_in_gb": 512,
        "scan_dns_record_id": "ocid1.vcndnsrecord.oc1.iad....q",
        "scan_ip_ids": null,
        "shape": "VM.Standard2.1",
        "sparse_diskgroup": null,
        "ssh_public_keys": [
        ],
        "subnet_id": "ocid1.subnet.oc1.iad....",
        "time_created": "2019-12-10T19:10:58.044000+00:00",
        "time_zone": "UTC",
        "version": "19.4.0.0.190716",
        "vip_ids": null
    },
    "invocation": {
        "module_args": {
            "api_user": null,
            "api_user_fingerprint": null,
            "api_user_key_file": null,
            "api_user_key_pass_phrase": null,
            "auth_type": "api_key",
            "availability_domain": null,
            "backup_subnet_id": null,
            "cluster_name": null,
            "compartment_id": null,
            "config_file_location": null,
            "config_profile_name": null,
            "cpu_core_count": null,
            "data_storage_percentage": null,
            "data_storage_size_in_gbs": 2048,
            "database_edition": null,
            "db_home": null,
            "db_system_id": "ocid1.dbsystem.oc1.iad.....a",
            "defined_tags": null,
            "delete_ssh_public_keys": false,
            "disk_redundancy": null,
            "display_name": null,
            "domain": null,
            "force_create": false,
            "freeform_tags": null,
            "hostname": null,
            "initial_data_storage_size_in_gb": null,
            "key_by": null,
            "license_model": null,
            "node_count": null,
            "purge_ssh_public_keys": true,
            "region": null,
            "shape": null,
            "source": "NONE",
            "ssh_public_keys": null,
            "state": "present",
            "subnet_id": null,
            "tenancy": null,
            "version": {
                "action": "PRECHECK",
                "patch_id": "ocid1.dbpatch.oc1.iad......"
            },
            "wait": false,
            "wait_timeout": 1200,
            "wait_until": null
        }
    }
}

**Comment by rishimahajan** [comment]
When I tried setting data_storage_size_in_gbs as 512 (as reported by error)
```
TASK [patch_db : Run PRECHECK for patch [u'Oct 2019 19c Db System patch']] ***********************************************************************************
Thursday 12 December 2019  03:29:56 +0000 (0:00:00.053)       0:00:07.015 ***** 
fatal: [localhost]: FAILED! => {"changed": false, "msg": "{'status': 400, 'message': u'The minimum Data storage size (in GB) for upgrade is 512.', 'code': u'InvalidParameter', 'opc-request-id': 'CAF5C7DA95DA45A1A6F0C633E4DB30D7/19ECCC79C04C2CEF8D19B80906746819/E80C0ED1847721E06A6CF832D63283E4'}"}
```
I got similar error for cpu_core_count .
```
TASK [patch_db : Run PRECHECK for patch [u'Oct 2019 18c Db System patch']] ***********************************************************************************
Thursday 12 December 2019  05:19:02 +0000 (0:00:00.050)       0:00:07.189 ***** 
fatal: [localhost]: FAILED! => {"changed": false, "msg": "{'status': 400, 'message': u'The maximum CPU core count for VM.Standard2.2 DB System is 2.', 'code': u'InvalidParameter', 'opc-request-id': '1043CD30E7F0490B92BE0668191BAE01/C4A761DAB0CF0BAB15E17726764F60EB/7866ACD8A6BD75BF3B17B346D913AF66'}"}
```
So I added both - data_storage_size_in_gbs and  cpu_core_count  in task and kicked off the precheck

DB system size initially was 256GB and shape was VM2.2 (RAC) with 4 OCPUs
```
- name: Run PRECHECK for patch {{ db_system_patch_facts | json_query('db_system_patches[].description') }}
  oci_db_system:
    db_system_id: "{{ db_system_ocid }}"
    data_storage_size_in_gbs: 512
    cpu_core_count: 2
    version:
      patch_id: "{{ db_system_patch_id }}"
      action: 'PRECHECK'
    state: 'present'
    wait: true
    wait_timeout: 3600
  when: db_system_patch_id|length > 0
  register: db_system_precheck_facts
```
The precheck operation went through and also updated the storage of the system from 256 to 512GB.
```
Before
Available Data Storage:256 GB
Total Storage Size (GB):912 GB

After
Available Data Storage:512 GB
Total Storage Size (GB):1168 GB
```
Now, a rerun of precheck operation is now asking for another upgrade to 1024GB
```
TASK [patch_db : Run PRECHECK for patch [u'Oct 2019 18c Db System patch']] ***********************************************************************************
Thursday 12 December 2019  06:01:02 +0000 (0:00:00.057)       0:00:07.208 ***** 
fatal: [localhost]: FAILED! => {"changed": false, "msg": "{'status': 400, 'message': u'The minimum Data storage size (in GB) for upgrade is 1024.', 'code': u'InvalidParameter', 'opc-request-id': '93682D80FDF6477FA303A6FDCF88180B/5D7719CD13D3B87F2D53A5B791C312EB/B8AC75FE0CE099B2B1E9DB2F72ACD54B'}"}
```

**Comment by nalsaber** [comment]
Fix is released in version [1.14.0](https://github.com/oracle/oci-ansible-modules/releases/tag/v1.14.0)
